torch/csrc/jit/pickler.cpp

```diff
@@ -0,0 +1,490 @@
+    return &(ivalue.toStringRef());
+  } else if (ivalue.isIntList()) {
+    return &(ivalue.toIntListRef());
+  }
+
+  return nullptr;
+}
+
+void Pickler::pushBinGet(uint32_t memo_id) {
+  if (memo_id <= std::numeric_limits<uint8_t>::max()) {
+    pushOpCode(OpCode::BINGET);
+    pushUint8(memo_id);
+  } else {
+    // Memoized too many items, issue a LONG_BINGET instead
+    pushOpCode(OpCode::LONG_BINGET);
+    pushUint32(memo_id);
+  }
+}
+
+void Pickler::pushString(const IValue& ivalue) {
+  const auto& string = ivalue.toStringRef();
+
+  pushOpCode(OpCode::BINUNICODE);
+  pushUint32(string.size());
+  pushString(string);
+  pushMemoization(ivalue);
+}
+
+void Pickler::pushString(const std::string& string) {
+  stack_.insert(stack_.end(), string.begin(), string.end());
+}
+
+void Pickler::pushClass(PicklerClass cls) {
+  const auto& name = getClassName(cls);
+  // Write it to the tensor table
+  auto memo_entry = memo_.find(&name);
+  if (memo_entry == memo_.end()) {
+    pushOpCode(OpCode::GLOBAL);
+    // Module name + "\n"
+    pushString(getModuleName());
+    // Class name + "\n"
+    pushString(name);
+    pushMemoization((void*)&name);
+  } else {
+    pushBinGet(memo_entry->second);
+  }
+
+  pushOpCode(OpCode::EMPTY_TUPLE);
+  pushOpCode(OpCode::NEWOBJ);
+}
+
+void Pickler::pushTensor(const IValue& ivalue) {
+  pushClass(PicklerClass::TENSOR);
+
+  tensor_table_.push_back(ivalue.toTensor());
+  auto tensor_id = tensor_table_.size() - 1;
+  pushOpCode(OpCode::BININT);
+  pushUint32(tensor_id);
+
+  pushOpCode(OpCode::BUILD);
+}
+
+void Pickler::pushIntList(const IValue& ivalue) {
+  pushClass(PicklerClass::INTLIST);
+
+  pushOpCode(OpCode::EMPTY_LIST);
+  pushMemoization(ivalue);
+  pushOpCode(OpCode::MARK);
+
+  for (const auto& item : ivalue.toIntListRef()) {
+    addIValue(item);
+  }
+
+  pushOpCode(OpCode::APPENDS);
+  pushOpCode(OpCode::BUILD);
+}
+
+void Pickler::pushDouble(const IValue& ivalue) {
+  double value = ivalue.toDouble();
+  AT_ASSERT(sizeof(double) == 8);
+  char* bytes = reinterpret_cast<char*>(&value);
+
+  pushOpCode(OpCode::BINFLOAT);
+  for (size_t i = 0; i < 8; ++i) {
+    pushUint8(bytes[8 - i - 1]);
+  }
+}
+
+using ivalue_pair = std::pair<IValue, IValue>;
+
+struct IValueComparator {
+  bool operator()(const ivalue_pair& lhs, const ivalue_pair& rhs) const {
+    if (lhs.first.isString()) {
+      return lhs.first.toStringRef() < rhs.first.toStringRef();
+    }
+    if (lhs.first.isInt()) {
+      return lhs.first.toInt() < rhs.first.toInt();
+    }
+    if (lhs.first.isDouble()) {
+      return lhs.first.toDouble() < rhs.first.toDouble();
+    }
+    AT_ERROR("Uncomparable IValue types");
+  }
+};
+
+void Pickler::pushDict(const IValue& ivalue) {
+  auto dict = ivalue.toGenericDictRef();
+
+  pushOpCode(OpCode::EMPTY_DICT);
+  pushMemoization(ivalue);
+
+  pushOpCode(OpCode::MARK);
+
+  // Sort the dict for deterministic keys
+  std::multiset<std::pair<IValue, IValue>, IValueComparator> dict_items(
```

Comment:
multiset is probably very very slow (std::set sucks), use vector of pairs and call sort() ?